### PR TITLE
Add 2 new topics

### PR DESCRIPTION
**Pull request by Vonfry**
Here may be some error of description.

### CODE REVIEW

**Comment by LEXUGE** [comment]
LGTM 
@linyinfeng

File: README.md [hunk 1]
```diff
@@ -116,3 +116,42 @@ nixpkgs=/nix/var/nix/profiles/per-user/root/channels/nixos:nixos-config=/etc/nix
...
+keep-outputs = true
+keep-derivations = true
+```
+
+它们可以避免在 GC 时删除你的 drv 和编译时依赖，从而减少打包时不必要的重复构建。更多说明参考文档 `man nix.conf` 。
+
+### 构建失败时保留临时文件
+`nix-build` 在构建时默认执行在临时目录（详细查询 `$NIX_BUILD_TOP` 相关文档）。临时目录通常是在 `/tmp` 中（详细查询 `TMPDIR, TEMPDIR, TMP, TEMP` 等）。 `--keep-failed` 选项可以在构建失败的时候，保留这些文件，这时你可以使用 `nix-shell` 并进行构建目录来排查错误。
+
+更进一步，nixpkgs 提供了 `breakpointHook` 可以让你在构建失败的时候产生提示信息来指导你后续调试，即使用 `cntr attach` 。
```
Comment (Ninlives) [inline README.md]: I think we should explain how to use breakpintHook here.

File: README.md [hunk 2]
```diff
@@ -116,3 +116,42 @@ nixpkgs=/nix/var/nix/profiles/per-user/root/channels/nixos:nixos-config=/etc/nix
...
+
+这种错误经常发生打包闭源二进制文件上，其中 elf 头的 interpreter 不正确上。可使用 `patchelf --set-interpreter` 进行修复，或者使用 nixpkgs 的 `autoPatchelfHook` 来自动完成。主要原理是 elf 的 interpreter 一般为 `/lib64/ld-linux-x86-64.so.2` ，但在 NixOS 中，这一目录是不存在的。我们需要修改为当前系统使用 glibc 、你所使用的其它 glibc 或相关库中的 ld。
+
+### Segment fault
+
+当确保编译过程没有问题的情况下，可以尝试关闭 strip ，避免移去不必要的内容。
+
+### jar 文件无法找到库
+
+类似于 elf 文件的 rpath 问题，处理方法有两种。一是解包 jar 文件，使用 `patchelf` 把 rpath 修正，再打包回去。二是直接打包 `LD_LIBRARY_PATH` 。
```
Comment (linyinfeng) [inline README.md]: Interesting case. What should be patched in the jar file by `patchelf`? Does this means there are elf executables or shared objects in the jar file?

File: README.md [hunk 3]
```diff
@@ -116,3 +116,53 @@ nixpkgs=/nix/var/nix/profiles/per-user/root/channels/nixos:nixos-config=/etc/nix
...
+<details><summary>3. Nix打包时常见错误</summary>
+<p>
+
+### no such file or directory
+
+这种错误经常发生打包闭源二进制文件上，其中 elf 头的 interpreter 不正确上。可使用 `patchelf --set-interpreter` 进行修复，或者使用 nixpkgs 的 `autoPatchelfHook` 来自动完成。主要原理是 elf 的 interpreter 一般为 `/lib64/ld-linux-x86-64.so.2` ，但在 NixOS 中，这一目录是不存在的。我们需要修改为当前系统使用 glibc 、你所使用的其它 glibc 或相关库中的 ld。
+
+### Segment fault
+
+当确保编译过程没有问题的情况下，可以尝试关闭 strip ，避免移去必要的内容。
```
Comment (Ninlives) [inline README.md]: We may add example here

```nix
# https://nixos.org/manual/nixpkgs/stable/#ssec-fixup-phase
stdenv.mkDerivation {
    # Other arguments
    dontStrip = true;
}
```

Everything else looks good to me, shall we merge now?

**Review by Ninlives — APPROVED** [review]

**Pull request merged**